### PR TITLE
feat(assign): refine types

### PR DESCRIPTION
**Pull request by crishoj**
## Summary

The type declaration with `assign` (mistakenly?) requires that both arguments have the same shape.

Taking the example from the docs:

image

Two warnings:

image

image

After:

image

## For any code change,

- [ ] Related documentation has been updated, if needed
- [ ] Related tests have been added or updated, if needed
- [ ] Related benchmarks have been added or updated, if needed

## Does this PR introduce a breaking change?

No

### CODE REVIEW

File: src/object/assign.ts [hunk 1]
```diff
@@ -14,10 +14,10 @@ import { isPlainObject } from 'radashi'
...
  * ```
  */
-export function assign<X extends Record<string | symbol | number, any>>(
-  initial: X,
-  override: X,
-): X {
+export function assign<
+  X extends Record<string | symbol | number, any>,
+  Y extends Record<string | symbol | number, any>,
+>(initial: X, override: Y): X & Y {
```
Comment (aleclarson) [inline src/object/assign.ts:20]: Unfortunately, `X & Y` is not representative.

```ts
type A = {a: number}
type B = {a: string}

type C = A & B

declare const c: C
c.a // number & string
```
[Link](https://typescript-play.js.org/#code/C4TwDgpgBAglC8UDeBDAXFAdgVwLYCMIAnAXwChRIoAhBZdKAZ2CIEtMBzci8aAYTpwAZDTJkAJhADGAGxRFoUgPaZmUKRj5kpAOhRQA9Aax5CRKCOZtOQA) to above example in playground.

We actually want `c.a` to result in `string` (as string will always override number). If `B["a"]` was possibly undefined like `{a?: string}`, then `c.a` should result in `number | string`.

**Review by aleclarson — CHANGES_REQUESTED** [review]
See comment :v:

**Comment by aleclarson** [comment]
Thanks for the contributions, @crishoj!

I agree the current `assign` types are not good :)

File: src/object/assign.ts (hunk 1 above)
Comment (crishoj) [reply src/object/assign.ts:20]: You're right, `A & B` does not result in the desired type.

Perhaps something like the `Merge` TS helper from https://github.com/millsp/ts-toolbelt/blob/master/sources/Object/Merge.ts could work?

**Comment by aleclarson** [comment]
Closed in favor of #142

**Pull request closed without merging**